### PR TITLE
Enable setting a fraction of the epoch as warmup

### LINKED ISSUE #1234: Add --warmup_fraction to pretraining script
Luca suggested to add `pretrain --warmup_fraction 0.05` to define warmup in terms of a fraction of the epoch, which we are doing in one of our Studios. It's just something I decided on my own based on the paper referenced there, whether this is widely used to do warmup this way I don't know.

There were apparently users who tried running the code outside the Studio but it failed, so we now need to support this argument in LitGPT.

### PR DESCRIPTION
**Pull request by awaelchli**
Fixes #1234 

This feature was requested by @lantiga through a customer. 

Context: I had this implemented in a Studio by modifying the LitGPT source. It is useful for continued pretraining where you want to perform warmup on a fraction of your dataset (say 0.05), so you could set `--train.lr_warmup_fraction=0.05`. This doesn't apply to regular pretraining where warmup is only a few thousand steps, and it wouldn't be very intuitive. Note that without this new arg, you would compute the steps in advance and set it as `lr_warmup_steps`.

### CODE REVIEW

File: litgpt/pretrain.py [hunk 1]
```diff
@@ -369,6 +370,14 @@ def get_dataloaders(
     return train_dataloader, val_dataloader
 
 
+def compute_warmup_iters(devices, max_iters, train, train_dataloader) -> int:
```
Comment (carmocca) [inline litgpt/pretrain.py:373]: Can we make this a method in `TrainArgs`? Similarly to how we have `train.gradient_accumulation_iters(devices)`

File: litgpt/pretrain.py [hunk 2]
```diff
@@ -426,6 +435,8 @@ def validate_args(train: TrainArgs, eval: EvalArgs, initial_checkpoint_dir, resu
                 issues.append(f"{__file__} requires the {name!r} argument. This is set in {args}")
     if initial_checkpoint_dir and resume:
         issues.append("Can't provide both `--resume` and `--initial_checkpoint_dir`. Choose one.")
+    if train.lr_warmup_fraction and train.lr_warmup_steps:
+        issues.append("Can't provide both `--train.lr_warmup_fraction` and `--train.lr_warmup_steps`. Choose one.")
```
Comment (carmocca) [inline litgpt/pretrain.py:439]: Can you do this validation in TrainArgs.__post_init__?

**Review by carmocca — APPROVED** [review]

File: litgpt/args.py [hunk 3]
```diff
@@ -16,8 +16,10 @@ class TrainArgs:
     """Number of samples between optimizer steps across data-parallel ranks"""
     micro_batch_size: int = 4
     """Number of samples per data-parallel rank"""
-    lr_warmup_steps: int = 100
+    lr_warmup_steps: Optional[int] = 100
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply litgpt/args.py:19]: I don't think so, because this is now the way to disable warmup if you want, no?
